alibi/explainers/partial_dependence.py

```diff
@@ -0,0 +1,1294 @@
+
+import matplotlib.pyplot as plt
+import numpy as np
+from alibi.api.defaults import DEFAULT_DATA_PD, DEFAULT_META_PD
+from alibi.api.interfaces import Explainer, Explanation
+from alibi.explainers.ale import get_quantiles
+from alibi.explainers.similarity.grad import get_options_string
+from sklearn.base import BaseEstimator, is_classifier, is_regressor
+from sklearn.ensemble import RandomForestRegressor
+from sklearn.ensemble._gb import BaseGradientBoosting
+from sklearn.ensemble._hist_gradient_boosting.gradient_boosting import \
+    BaseHistGradientBoosting
+from sklearn.inspection._partial_dependence import (
+    _grid_from_X, _partial_dependence_brute, _partial_dependence_recursion)
+from sklearn.tree import DecisionTreeRegressor
+from sklearn.utils import _get_column_indices, _safe_indexing
+from sklearn.utils.extmath import cartesian
+from sklearn.utils.validation import check_is_fitted
+
+logger = logging.getLogger(__name__)
+
+
+class ResponseMethod(str, Enum):
+    """ Enumeration of supported response_method. """
+    AUTO = 'auto'
+    PREDICT_PROBA = 'predict_proba'
+    DECISION_FUNCTION = 'decision_function'
+
+
+class Method(str, Enum):
+    """ Enumeration of supported method. """
+    AUTO = 'auto'
+    RECURSION = 'recursion'
+    BRUTE = 'brute'
+
+
+class Kind(str, Enum):
+    """ Enumeration of supported kind. """
+    AVERAGE = 'average'
+    INDIVIDUAL = 'individual'
+    BOTH = 'both'
+
+
+class PartialDependence(Explainer):
+    def __init__(self,
+                 predictor: Union[BaseEstimator, Callable[[np.ndarray], np.ndarray]],
+                 feature_names: Optional[List[str]] = None,
+                 categorical_names: Optional[Dict[int, List[str]]] = None,
+                 target_names: Optional[List[str]] = None,
+                 predictor_kw: Optional[Dict[str, Any]] = None):
+        """
+        Partial dependence for tabular datasets. Supports one feature or two feature interactions.
+
+        Parameters
+        ----------
+        predictor
+            A `sklearn` estimator or a prediction function which receives as input a `numpy` array and outputs a
+            `numpy` array.
+        feature_names
+            A list of feature names used for displaying results.
+        categorical_names
+            Dictionary where keys are feature columns and values are the categories for the feature. Necessary to
+            identify the categorical features in the dataset.
+        target_names
+            A list of target/output names used for displaying results.
+        predictor_kw
+            Predictor identifier arguments when the predictor is a callable prediction function. The following
+            arguments must be provided:
+
+             - ``'predictor_type'`` : ``str`` - Type of the predictor. Available
+                values: ``'regressor'`` | ``'classifier'``.
+
+             - ``'prediction_fn'`` : ``str`` - Name of the prediction function. Available value for
```

Comment:
It is true that we just need to call the function, but unfortunately the `sklearn` backend treats them a bit different in some cases. For example, in the binary classification setting, the `sklearn` remove the the proabability corresponding to the class 0. On the other hand, we decided offline that we should include that too as it would be consistent with the ALE. Although for probability we have the 'liberty' to choose whether we include or not include the probability for class 0, for decision function there is only one option. For a binary classifier, there is a single decision function. Hence the special treatment based on the type of predictor (i.e., `decision_function` or `probability`). Happy to discuss alternatives offline.